lib/propolis/src/hw/qemu/fwcfg.rs

```diff
@@ -396,259 +304,353 @@ impl FwCfg {
-            FwCfgDmaCtrl::READ | FwCfgDmaCtrl::WRITE | FwCfgDmaCtrl::SKIP,
-        ) {
-            state.offset.checked_add(dma_req.len).ok_or("offset overflow")?
+        let len = if let Some(buf) = selected.cached_value.as_ref() {
+            write_buf(buf, &mut ro)
         } else {
-            state.offset
+            match entry {
+                Entry::RamFb => {
+                    if let Some(ramfb) = state.ramfb.as_ref() {
+                        ramfb
+                            .fwcfg_rw(RWOp::Read(&mut ro))
+                            .map_err(|_| FwCfgErr::OpUnsuccessful)?;
+                        ro.bytes_written()
+                    } else {
+                        0
+                    }
+                }
+                Entry::Bytes(buf) => write_buf(&buf, &mut ro),
+                Entry::FileDir => {
+                    panic!("expected intact cached buffer for static entry");
+                }
+            }
         };
+        selected.offset = selected
+            .offset
+            .checked_add(len as u32)
+            .expect("offset does not overflow");
+        Ok(len)
+    }
 
-        // match the same command precedence as qemu (read, write, skip)
-        if ctrl.contains(FwCfgDmaCtrl::READ) {
-            let res = self.dma_read(
-                state.selector,
-                dma_req.addr,
-                state.offset,
-                dma_req.len,
-                mem,
-            );
-            if res.is_err() {
-                return res;
-            }
-        } else if ctrl.contains(FwCfgDmaCtrl::WRITE) {
-            let res = self.dma_write(
-                state.selector,
-                dma_req.addr,
-                state.offset,
-                dma_req.len,
-                mem,
-            );
-            if res.is_err() {
-                return res;
+    fn write(
+        &self,
+        state: &mut MutexGuard<State>,
+        wo: &mut WriteOp,
+    ) -> Result<usize, FwCfgErr> {
+        let state = state.deref_mut();
+        let selected = state.selected.as_mut().ok_or(FwCfgErr::NoneSelected)?;
+        let entry = state
+            .directory
+            .entry(selected.selector)
+            .expect("selected entry is present");
+        // Encode the current offset into the WriteOp (as a child)
+        let mut wo = WriteOp::new_child(selected.offset as usize, wo, ..);
+
+        let len = match entry {
+            Entry::FileDir | Entry::Bytes(_) => Err(FwCfgErr::ReadOnly),
+            Entry::RamFb => {
+                if let Some(ramfb) = state.ramfb.as_ref() {
+                    ramfb
+                        .fwcfg_rw(RWOp::Write(&mut wo))
+                        .map_err(|_| FwCfgErr::OpUnsuccessful)?;
+                    Ok(wo.bytes_read())
+                } else {
+                    Ok(0)
+                }
             }
-        }
-        state.offset = end_offset;
-        Ok(())
+        }?;
```

Comment:
this is, admittedly, a very minor nitpick, but i find the `Ok`-wrapping here a little bit unnecessarily convoluted --- perhaps this would be a bit clearer:
```rust
let len = match entry {
    Entry::FileDir | Entry::Bytes(_) => return Err(FwCfgErr::ReadOnly),
    Entry::RamFb => {
       if let Some(ramfb) = state.ramfb.as_ref() {
            ramfb
                .fwcfg_rw(RWOp::Write(&mut wo))
                .map_err(|_| FwCfgErr::OpUnsuccessful)?;
            wo.bytes_read()
        } else {
            0
        }
    }
};
```

Reply:
I thought it was clearer with the single bail-out point, so I'm inclined to leave it.